### PR TITLE
POC of simple instantiate (resolve then rec instantiate)

### PR DESCRIPTION
**Pull request by guillaumegenthial**
## Motivation

Another attempt to solve https://github.com/facebookresearch/hydra/issues/1183

This proposal involves a major change in the instantiation logic.

DO NOT MERGE

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes/No

## Test Plan

Run this test

```bash
pytest tests/test_utils.py::test_instantiate_dataclass_modes --capture=no
```

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: hydra/utils.py [hunk 1]
```diff
@@ -1,36 +1,34 @@
...
+class ConvertMode(str, Enum):
     # Use DictConfig/ListConfig
-    NONE = 0
+    NONE = "none"
     # Convert the OmegaConf config to primitive container, Structured Configs are preserved
-    PARTIAL = 1
+    PARTIAL = "partial"
     # Fully convert the OmegaConf config to primitive containers (dict, list and primitives).
-    ALL = 2
+    ALL = "all"
```
Comment (omry) [inline hydra/utils.py:31]: curious about this one. can you tell me why you changed this?

Comment (guillaumegenthial) [reply hydra/utils.py:31]: ah yes, just a simple trick I discovered recently to simplify the use of Enum (see https://docs.python.org/3/library/enum.html#others).

What you get from doing this is
1. equality : `ConvertMode.NONE == "none"` is now true
2. resolution `ConvertMode("none")` returns `ConvertMode.NONE`

Comment (omry) [reply hydra/utils.py:31]: Interesting. OmegaConf is handling some of it for configs. I was not aware of this trick.

```
from dataclasses import *
from enum import Enum
from omegaconf import *

class Color(Enum):
    red = 1
    green = 2
    blue = 3

class StrColor(str, Enum):
    red = "red"
    green = "green"
    blue = "blue"


@dataclass
class Foo:
    color: Color = Color.red
    scolor: StrColor = StrColor.red

cfg : Foo = OmegaConf.structured(Foo)
print("cfg", cfg)

cfg.color = Color.blue
print("cfg", cfg)

# automatic conversion works
cfg.color = "green" # mypy complains
print("cfg", cfg)

# comparission does not work
assert cfg.color != "green"

## str color
print("str color")
cfg.scolor = "green" # mypy still complains
print("cfg", cfg)

# equality with strings works
assert cfg.scolor == "green"
```

**Comment by guillaumegenthial** [comment]
Simple POC, closing

**Pull request closed without merging**